handler/handler.go

```diff
@@ -2,47 +2,206 @@ package handler
-		title := fmt.Sprintf("SearchDataImportModel.Title, %s!", events[i].Title)
-		keywords := fmt.Sprintf("SearchDataImportModel.Keywords, %s!", events[i].Keywords)
-		traceId := fmt.Sprintf("SearchDataImportModel.TraceId, %s!", events[i].TraceID)
-
-		logData := log.Data{
-			"dataType":    dataType,
-			"searchIndex": searchIndex,
-			"title":       title,
-			"keywords":    keywords,
-			"traceId":     traceId,
-		}
-		log.Info(ctx, "Event in Pipeline to be processed for Elastic Search", log.INFO, logData)
+	//no events receeived. Nothing more to do.
+	if len(events) == 0 {
+		log.Info(ctx, "there are no events to handle")
+		return nil
+	}
+	go status(ctx)
+
+	// This will block if we've reached our concurrency limit (sem buffer size)
+	err := batchHandler.SendToES(ctx, events, len(events))
+	if err != nil {
+		log.Fatal(ctx, "failed to send event to Elastic Search", err)
+		return err
 	}
 
+	time.Sleep(4 * time.Second)
+	syncWaitGroup.Wait()
+
 	log.Info(ctx, "event successfully handled")
 	return nil
 }
+
+func (bh BatchHandler) SendToES(ctx context.Context, events []*models.SearchDataImportModel, length int) error {
+
+	// Wait on semaphore if we've reached our concurrency limit
+	syncWaitGroup.Add(1)
+	semaphore <- 1
+
+	// Get Config
+	cfg, err := config.Get()
+	if err != nil {
+		log.Fatal(ctx, "error getting config", err)
+		os.Exit(1)
+	}
+
+	esDestURL := cfg.ElasticSearchAPIURL
+	esIndex := "esIndex"
+	esDestType := "docType"
+
+	esDestIndex := fmt.Sprintf("%s/%s", esIndex, esDestType)
+	log.Info(ctx, "esDestIndex ", log.Data{"esDestIndex": esDestIndex})
+
+	t := transform.NewTransformer()
+
+	go func() {
+		defer func() {
+			<-semaphore
+			syncWaitGroup.Done()
+		}()
+
+		countChannel <- length
+		target := length
+		var bulk []byte
+
+		i := 0
+		for i < length {
+
+			if events[i].Title == "" {
+				log.Info(ctx, "No title for inbound event, no transformation possible")
+				continue // break here
+			}
+
+			esmodel := t.TransformEventModelToEsModel(events[i])
+
+			if esmodel != nil {
+				b, err := json.Marshal(esmodel)
+				if err != nil {
+					log.Fatal(ctx, "error marshal to json", err)
+					return
+				}
+
+				bulk = append(bulk, []byte("{ \"create\": { \"_id\": \""+esmodel.Title+"\" } }\n")...)
+				bulk = append(bulk, b...)
+				bulk = append(bulk, []byte("\n")...)
+			} else {
+				skipChannel <- 1
+				target--
+			}
+			i++
+		}
+
+		b, err := bh.ElasticSearchCli.SubmitBulkToES(ctx, esDestIndex, esDestURL, bulk)
+		if err != nil {
+			log.Fatal(ctx, "error in submitting bulk in ES", err)
+			return
+		}
+
+		var bulkRes esBulkResponse
+		if err := json.Unmarshal(b, &bulkRes); err != nil {
+			log.Info(ctx, "error unmarshaling json", log.Data{
+				"actual response": b,
+				"err":             err,
+			})
+			log.Error(ctx, "error unmarshaling json", err)
+		}
+
+		if bulkRes.Errors {
+			for _, r := range bulkRes.Items {
+				if r["create"].Status != 201 {
+					log.Fatal(ctx, "error inserting doc to ES", err)
```

Comment:
This should be log.Error; fatal is reserved for killing off application and we wont be doing this here

Please add a TODO comment to handle failed updates to elasticsearch. The plan will be to requeue these failed events